Implement Fully Sharded Data Parallel (FSDP) in PyTorch XLA

ronghanghu opened the pull request:
This PR implements Fully Sharded Data Parallel (FSDP) in PyTorch XLA.

Example usage:
```
my_module = my_module.to(xm.xla_device())
sharded_module = XlaFullyShardedDataParallel(my_module)
optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
x = sharded_module(x, y=3, z=torch.Tensor([1]))
loss = x.sum()
loss.backward()
optim.step()
```
Example training scripts:
* MNIST: `test/test_train_mp_mnist_fsdp_with_ckpt.py` (it also tests checkpoint consolidation)
* ImageNet: `test/test_train_mp_imagenet_fsdp.py`

Notes:
* The `XlaFullyShardedDataParallel` class supports both the ZeRO-2 optimizer (sharding gradients and optimizer states) and the ZeRO-3 optimizer (sharding parameters, gradients, and optimizer states) in https://arxiv.org/abs/1910.02054. The ZeRO-3 optimizer should be implemented via nested FSDP. See `test/test_train_mp_mnist_fsdp_with_ckpt.py` for an example.
* `flatten_parameters` (default to `True`) will flatten all parameters in a module into a single tensor for sharding. This is usually more efficient compared to sharding each individual parameter.
* when saving checkpoints during training, also save `model.get_shard_metadata()` along with `model.state_dict()` and use `consolidate_xla_fsdp_model_checkpoints` to stitch the sharded model checkpoints together into a full model state dict. See `test/test_train_mp_mnist_fsdp_with_ckpt.py` for an example. The script `checkpoint_consolidation.py` can also be launched from the command line.
* The old `all_reduce` implementation from PyTorch XLA 1.10 is used by default to avoid the GRPC error (see https://github.com/pytorch/xla/issues/3423).
* The implementation of this class is largely inspired by and mostly follows the structure of `fairscale.nn.FullyShardedDataParallel` in https://fairscale.readthedocs.io/en/stable/api/nn/fsdp.html.

---

TODOs
- [ ] figure out a good place to put the `fsdp` module (currently its under `test/` but maybe we should move it to `torch_xla/utils` or `torch_xla/distributed`).

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+      p._sharded_param = p_shard  # add a handle to the sharded parameter
+      # Free the full parameter storage (here we free its `.data`) but keep the tensor itself
+      # for auto-grad tracing (like `torch.autograd.Variable` before the tensor-variable merge).
+      p.data = self._dummy_data_placeholder
```
So I'm trying to free the memory buffer by assigning its `.data` to a dummy tensor of size 1.

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+      self,
+      module: nn.Module,
+      reshard_after_forward: bool = True,
+      flatten_parameters: bool = True,
```
Thanks for the comments! Could you elaborate on why the first situation will cause MPMD? It's my understanding that when sharding flattened parameters, everything will still be symmetric on each worker (for example, all workers will do the same optimizer update, etc) and there aren't any worker-specific operations compared to the case of sharding unflattened individual parameters, so SPMD will still work.

hjm-aws replied:
Thanks Ronghang for the explanation. Previously I didn't know about Step 4, that is `backword()` on a view can generate `.grad` on the backing tensor.

hjm-aws commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+    shared_full_param_memo = {}
+    shared_full_param_infos = []
+    full_params = []
+    for module_name, m in self.named_modules():
+      for n, p in m.named_parameters(recurse=False):
```
Will these two lines yield `self._fsdp_wrapped_module.flat_params`? I failed to see how it could by looking at https://github.com/pytorch/pytorch/blob/v1.11.0/torch/nn/modules/module.py#L1668.

ronghanghu replied:
Yes, these two lines will yield `self._fsdp_wrapped_module.flat_params`. In `XlaFlattenParamsWrapper` the original parameters of the module have been replaced by the flattened parameters.

hjm-aws replied:
Sure, but how is the original `module` being accessed through `self.named_modules()`?

hjm-aws replied:
Sure, but how is `self._fsdp_wrapped_module` accessed through `self.named_modules()`?

hjm-aws replied:
OK I think it's done through https://github.com/pytorch/pytorch/blob/v1.11.0/torch/nn/modules/module.py#L1217. NVM then. Thanks!

miladm commented:
I successfully ran FSDP to train `ResNet` on `TPUv3`. 
Reviewing the code.

hjm-aws commented:
@ronghanghu Hi Ronghang, just to confirm I read the code correctly: we don't have prefetch logic here, right? By prefetching I mean pre-`all-gather` the params for the next layer/submodule.

ronghanghu commented:
> @ronghanghu Hi Ronghang, just to confirm I read the code correctly: we don't have prefetch logic here, right? By prefetching I mean pre-`all-gather` the params for the next layer/submodule.

@hjm-aws You are right, we don't have a prefetch logic here. The all-gather of the next layer happens after the previous layer is completed (instead of happening in parallel). This is consistent with FairScale's FSDP implementation.

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+    if execute_sharding_on_init:
+      # Execute the parameter sharding immediately and free up the memory
+      xm.mark_step()
+      gc.collect()
```
(thanks! I changed to GC before mark_step)

ronghanghu commented:
As a temporary workaround to https://github.com/pytorch/xla/issues/3455#issuecomment-1085448513, I added a `mark_step_on_freeing` parameter in the `__init__`. If it is `True`, we call `xm.mark_step` immediately upon freeing full parameters.

This is a temporary and inefficient workaround to avoid XLA compiler fusion that breaks parameter freeing in nested FSDP (ZeRO-3). It is useful only when `reshard_after_forward` is `True`. See details in https://github.com/pytorch/xla/issues/3455#issuecomment-1085448513.

This workaround notably increases the execution time and may trigger more compilation and does not cover all the cases, so we need a permanent solution to #3455. I'll remove the `mark_step_on_freeing` option when #3455 is fixed.

hjm-aws commented:
@ronghanghu @JackCaoG Hi Ronghang and Jack, when trying out the PR with the following code on our end:
```
    fsdp_wrap = lambda m: FSDP(m.to(device), flatten_parameters=True)
  
    for submodule_name in [n for n, _ in model.named_children()]:
      m_fsdp = fsdp_wrap(getattr(model, submodule_name))
      setattr(model, submodule_name, m_fsdp)
      
    model = fsdp_wrap(model)
```

I am seeing the following failure:
```
  File "bert_pretrain_fsdp.py", line 137, in bert_train
    m_fsdp = fsdp_wrap(getattr(model, submodule_name))
  File "bert_pretrain_fsdp.py", line 132, in <lambda>
    fsdp_wrap = lambda m: FSDP(m.to(device), flatten_parameters=True)
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py", line 210, in __init__
    self._shard_parameters_(params_to_shard)
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py", line 423, in _shard_parameters_
    p.data = self._dummy_data_placeholder
RuntimeError: torch_xla/csrc/tensor.cpp:802 : Check failed: xla::util::Multiply<xla::int64>(ir_value.shape().dimensions()) == xla::util::Multiply<xla::int64>(view->shape().dimensions()) (1 vs. 96386048)
```
The CHECK failure should be at the following location:
https://github.com/pytorch/xla/blob/4037ca7f2febe0de03edfcd69a871ef5668b488f/torch_xla/csrc/tensor.cpp#L813

We printed out the tensor that was involved in this failure, and it was `_fsdp_wrapped_module.flat_param_0`.

I am confused by this failure. It looks as if `_fsdp_wrapped_module.flat_param_0` was created as a view, thus the failure of trying to assign a 1 element tensor to the view. But the code in this PR shows that `_fsdp_wrapped_module.flat_param_0` was created with `torch.cat`, which should not be creating views. Thoughts?

ronghanghu commented:
Hi @hjm-aws, you are probably running into the issue in https://github.com/pytorch/xla/issues/3392 (see the "Additional context" section) or https://github.com/pytorch/xla/issues/3502.

My solution (for today) is to install the following builds for TPU usage
```
# torch, torchvision and torch_xla 20220419
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-nightly+20220419-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torchvision-nightly+20220419-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-nightly+20220419-cp38-cp38-linux_x86_64.whl

# libtpu 20220408
sudo pip3 install https://storage.googleapis.com/cloud-tpu-tpuvm-artifacts/wheels/libtpu-nightly/libtpu_nightly-0.1.dev20220408-py3-none-any.whl
```

But as mentioned in https://github.com/pytorch/xla/pull/3507, TensorFlow version will be updated in tomorrow's 20220421 build, so one can also wait and try it tomorrow.

I'm sometimes observing higher TPU memory usage than expected when using this PR myself, so it's not finalized yet. My latest development is in https://github.com/ronghanghu/xla/tree/xla_fsdp_mem_dev and I'll update this PR one it's more finalized.

ronghanghu commented:
> I am confused by this failure. It looks as if `_fsdp_wrapped_module.flat_param_0` was created as a view, thus the failure of trying to assign a 1 element tensor to the view. But the code in this PR shows that `_fsdp_wrapped_module.flat_param_0` was created with `torch.cat`, which should not be creating views. Thoughts?

The `_fsdp_wrapped_module.flat_param_0` is created as a view of the all-gather output during full-parameter rebuilding, as in https://github.com/pytorch/xla/blob/1664252f66c29f93dfb00eb57d70e90ff44b07ef/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L889

hjm-aws commented:
Thanks Ronghang! It's very useful.

> The _fsdp_wrapped_module.flat_param_0 is created as a view of the all-gather output

The failure stack shows it was during the construction of FSDP wrapper, not after all-gather. That's why I am confused. Thoughts?

ronghanghu commented:
> Thanks Ronghang! It's very useful.
> 
> > The `_fsdp_wrapped_module.flat_param_0` is created as a view of the all-gather output
> 
> The failure stack shows it was during the construction of FSDP wrapper, not after all-gather. That's why I am confused. Thoughts?

I think it's because previously in `_shard_parameters_` a view was created on top of `_fsdp_wrapped_module.flat_param_0` via `_get_shard` as in https://github.com/pytorch/xla/blob/1664252f66c29f93dfb00eb57d70e90ff44b07ef/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L457 (here both flatten and chunk will create views).

So once a view is created on top of a tensor `t`, then one cannot assign its `t.data` to another tensor with a different shape before https://github.com/pytorch/xla/pull/3411 is merged (see the additional context in https://github.com/pytorch/xla/issues/3392). So you'll need a newer PyTorch XLA build that contains the fix https://github.com/pytorch/xla/pull/3411.

miladm reviewed (commented):
Thanks for this implementation @ronghanghu!

Added a few comments below. Feel free to rerun the tests so we can get them passed and prepare for next steps.

ronghanghu commented:
@miladm Thanks for the comments! Below are a few updates on my end for this PR:

I've been testing this PR myself on a transformer model wit with 10-billion parameters this week. So far it seems to be working OK, but I'm still debugging a few problems

1) It's weird that I had to insert an extra `xm.mark_step` call between model backward pass and AdamW optimizer stepping. Otherwise, the XLA compiler does not free the memory in time on v3-128 and causes OOM, but this problem doesn't happen on v3-8. I've tried inserting an `xm.optimization_barrier_` call to every input to the optimizer and still cannot get the same behavior as an explicit mark-step call. This behavior seems specifically related to all_gather -- If I replace all_gather with a dummy repeat op it doesn't happen.

This issue seems quite hard to isolate and I haven't managed to get a minimal reproducing example on it. I plan to submit a GitHub issue if I make more progress on this.

2) It seems that after I push my model to the XLA device via `model = model.to(device)`, it is still taking a large chunk of host CPU memory on the VM -- somehow the CPU tensors aren't freed in time.

This sometimes causes me a CPU OOM issue when constructing large models and sending their shards to TPUs. This problem is more pronounced when using `flatten_parameters=True`. I'm trying to see if I can get a minimal reproducing example for this. 

3) There seems to be a larger overhead on the TPU pod compared to v3-8. I'm trying to look into it but currently cannot generate a TPU device trace due to https://github.com/pytorch/xla/issues/3446 so trying to look at it by other means. (It would be great if the XLA profiler can run on the TPU pod as well, but so far this issue is not blocking me.)

I'll address your comments after resolving the issues above.

miladm commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1360 @@
+    else:
+      self.reduce_scatter_op = functools.partial(
+          xm.reduce_scatter, pin_layout=_pin_layout_in_reduce_scatter)
+    # TODO (ronghanghu): remove when https://github.com/pytorch/xla/issues/3455 is resolved
```
This issue is resolved at this point. Shall we handle this TODO prior to merge?

ronghanghu replied:
(this obsolete option is now removed)

miladm commented:
@ronghanghu the CI seems to been giving us trouble recently. I just hit rerun to let it get a chance to pass. This implementation is in Python. Technically, build should pass.

miladm commented:
@ronghanghu re: this [comment](https://github.com/pytorch/xla/pull/3431#issuecomment-1112358933) what model are your trying?

ronghanghu commented:
> @ronghanghu re: this [comment](https://github.com/pytorch/xla/pull/3431#issuecomment-1112358933) what model are your trying?

Hi, it's a [ViT](https://arxiv.org/abs/2010.11929) model with 10-billion parameters (using embed dim 5120, 32 attention heads, 32 layers, and an MLP ratio of 4).

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1030 @@
+      execute_sharding_on_init (bool, Optional):
+          if ``True``, immediately execute the parameter sharding via
+          `xm.mark_step` to free up the memory of the full parameters.
+      use_all_gather_via_all_reduce (bool, Optional):
```
(this obsolete arg is now removed)

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1360 @@
+    params = []
+    for param_name, param in module.named_parameters():
+      if not hasattr(param, "_is_sharded"):
+        param_names.append(param_name)
```
(removed)

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1360 @@
+      xm.wait_device_ops()
+
+  def _get_gradient_predivide_factor(self, world_size: int) -> float:
+    factor: int = 1
```
There is an explicit method `set_gradient_divide_factors` to allow users to override the factors later, so I'd like to keep it as-is.

miladm commented:
@ronghanghu I understand we have a few blockers to get `ViT` to run performantly on TPU. Though, it looks like those issues are being addressed separately. 

We'd like to prepare for merging our release 1.12 PRs as soon as possible. What are the main blockers for merging this PR in your view?

ronghanghu commented:
> @ronghanghu I understand we have a few blockers to get `ViT` to run performantly on TPU. Though, it looks like those issues are being addressed separately.
> 
> We'd like to prepare for merging our release 1.12 PRs as soon as possible. What are the main blockers for merging this PR in your view?

Hi @miladm, I think while this PR is working well on MNIST and ImageNet examples, I would prefer to first get more profiling results on large ViT models in computer vision or GPT models in NLP (which is exactly what this PR is intended to be applied to) before finalizing it.

- I think one main blocker is that, depending on the speed analyses and fixes (e.g. #3441), we might want to add or remove or change the default arguments in the `XlaFullyShardedDataParallel `'s `__init__` to give users more options (e.g. how and where `xm.optimization_barrier_` is applied, whether to overlay compute ops with collective ops). So the interface and the default arguments are not finalized yet and might need to be changed in a backward-incompatible way later.

- Another main blocker is #3545, which prevents this PR from being applied to larger models (e.g. 25 or 50 billion parameters) due to CPU OOM, although it could work on 10-billion scale models that cannot be trained without this PR.

Nevertheless, if we decide it is better to first merge it into the 1.12 release (and potentially modify it later with possible backward-incompatibility), then I can first address a few pending comments above and do some further clean-up and documentation.

miladm commented:
Thanks for the updates @ronghanghu. A few comments below:

+ Let's aim to make this code be part of release `1.12`. We can always make changes to it afterwards. 
+ We are aiming to do the branch cut in mid May. So, we have next week for merging the PR and until then it's possible we identify some resolution to the bugs you pointed out above.

As a next step, I suggest we aim to merge the PR by 5/15 at the latest. Wdyt?

ronghanghu commented:
> Thanks for the updates @ronghanghu. A few comments below:
> 
> * Let's aim to make this code be part of release `1.12`. We can always make changes to it afterwards.
> * We are aiming to do the branch cut in mid May. So, we have next week for merging the PR and until then it's possible we identify some resolution to the bugs you pointed out above.
> 
> As a next step, I suggest we aim to merge the PR by 5/13 at the latest. Wdyt?

@miladm Yeah, this sounds good to me. Then I can address the reviewer comments above and do more tests in the upcoming week.

ronghanghu commented:
## Tests to be performed before merging (05/06/2022)

@miladm I've done some further cleanup of this PR and addressed the reviewer comments above. The following tests will be performed to verify this FSDP implementation before we merge this PR into PyTorch XLA 1.12 release. I'll get them ready by next week. Thanks for your help!

#### Summary

- [ ] Tests on v3-8
  - [ ] Test MNIST FSDP ZeRO-2 on v3-8
  - [ ] Test MNIST FSDP ZeRO-2 + flatten params on v3-8
  - [ ] Test MNIST FSDP ZeRO-3 on v3-8
  - [ ] Test MNIST FSDP ZeRO-2 + flatten params on v3-8
  - [ ] Test MNIST FSDP ZeRO-3 + gradient checkpointing (and manual checkpoint consolidation) on v3-8
  - [ ] Test ImageNet ResNet-50 FSDP ZeRO-2 on v3-8
  - [ ] Test ImageNet ResNet-50 FSDP ZeRO-3 on v3-8
  - [ ] Test ImageNet ResNet-50 FSDP ZeRO-3 + flattened params + gradient checkpointing on v3-8
- [ ] Tests on v3-128 pod
  - [ ] Test ImageNet ResNet-50 FSDP ZeRO-2 on v3-128
  - [ ] Test ImageNet ResNet-50 FSDP ZeRO-3 + gradient checkpointing on v3-128
  - [ ] Test ViT 10-billion model on v3-128
  - [ ] (internal scaling tests at Meta AI)

### Tests on v3-8

#### Set up environment

```bash
# torch, torchvision and torch_xla 20220505
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-nightly+20220505-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torchvision-nightly+20220505-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-nightly+20220505-cp38-cp38-linux_x86_64.whl

# libtpu 20220413
sudo pip3 install https://storage.googleapis.com/cloud-tpu-tpuvm-artifacts/wheels/libtpu-nightly/libtpu_nightly-0.1.dev20220413-py3-none-any.whl

# install the FSDP PR into PyTorch/XLA
cd ~ && rm -rf xla_fsdp_dev && git clone https://github.com/ronghanghu/xla.git xla_fsdp_dev
cd xla_fsdp_dev && git checkout xla_fsdp_rebased
sudo rm -rf /usr/local/lib/python3.8/dist-packages/torch_xla/distributed/fsdp
sudo cp -r ./torch_xla/distributed/fsdp /usr/local/lib/python3.8/dist-packages/torch_xla/distributed/
cd ~
# for development using ~/workspace/xla_fsdp_clean_up
~/workspace/add_fsdp_to_torch_xla.sh
```

#### Test MNIST FSDP ZeRO-2 on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2
```

**Results**: (pending)

#### Test MNIST FSDP ZeRO-2 + flatten params on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --flatten_parameters
```

**Results**: (pending)

#### Test MNIST FSDP ZeRO-3 on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --use_nested_fsdp
```

**Results**: (pending)

#### Test MNIST FSDP ZeRO-3 + flattened params on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --use_nested_fsdp --flatten_parameters
```

**Results**: (pending)

#### Test MNIST FSDP ZeRO-3 + gradient checkpointing (and manual checkpoint consolidation) on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --use_nested_fsdp --use_gradient_checkpointing

python3 -m torch_xla.distributed.fsdp.consolidate_sharded_ckpts \
  --ckpt_prefix /tmp/mnist-fsdp/final_ckpt \
  --ckpt_suffix "_rank-*-of-*.pth"
```

**Results**: (pending)

#### Test ImageNet ResNet-50 FSDP ZeRO-2 on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_imagenet_fsdp.py \
  --datadir /datasets/imagenet-1k --drop_last \
  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
  --lr 0.4 --batch_size 128 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100
```

**Results**: (pending)

#### Test ImageNet ResNet-50 FSDP ZeRO-3 on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_imagenet_fsdp.py \
  --datadir /datasets/imagenet-1k --drop_last \
  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
  --lr 0.4 --batch_size 128 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100 \
  --use_nested_fsdp
```

**Results**: (pending)

#### Test ImageNet ResNet-50 FSDP ZeRO-3 + flattened params + gradient checkpointing on v3-8

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_imagenet_fsdp.py \
  --datadir /datasets/imagenet-1k --drop_last \
  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
  --lr 0.4 --batch_size 128 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100 \
  --use_nested_fsdp --flatten_parameters --use_gradient_checkpointing
```

**Results**: (pending)

### Tests on v3-128 pod

#### Set up environment

```bash
TPU_NAME=rh-128-0  # change to your TPU name
ZONE=europe-west4-a  # change to your TPU zone

gcloud alpha compute tpus tpu-vm ssh ${TPU_NAME} --zone ${ZONE} \
  --worker all \
  --command "
# torch, torchvision and torch_xla 20220505
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-nightly+20220505-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torchvision-nightly+20220505-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-nightly+20220505-cp38-cp38-linux_x86_64.whl

# libtpu 20220413
sudo pip3 install https://storage.googleapis.com/cloud-tpu-tpuvm-artifacts/wheels/libtpu-nightly/libtpu_nightly-0.1.dev20220413-py3-none-any.whl

# ViT dependency
sudo pip3 install timm==0.4.12

# install the FSDP PR into PyTorch/XLA
cd ~ && rm -rf xla_fsdp_dev && git clone https://github.com/ronghanghu/xla.git xla_fsdp_dev
cd xla_fsdp_dev && git checkout 20220505
sudo rm -rf /usr/local/lib/python3.8/dist-packages/torch_xla/distributed/fsdp
sudo cp -r ./torch_xla/distributed/fsdp /usr/local/lib/python3.8/dist-packages/torch_xla/distributed/

# clone this repo ViT FSDP example
cd ~ && rm -rf vit_10b_fsdp_example && git clone https://github.com/ronghanghu/vit_10b_fsdp_example.git
"
```

#### Test ImageNet ResNet-50 FSDP ZeRO-2 on v3-128

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_imagenet_fsdp.py \
  --datadir /datasets/imagenet-1k --drop_last \
  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
  --lr 3.2 --batch_size 64 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100
```

**Results**: (pending)

#### Test ImageNet ResNet-50 FSDP ZeRO-3 + gradient checkpointing on v3-128

```bash
python3 -u ~/xla_fsdp_dev/test/test_train_mp_imagenet_fsdp.py \
  --datadir /datasets/imagenet-1k --drop_last \
  --model resnet50 --test_set_batch_size 64 --eval_interval 10 \
  --lr 3.2 --batch_size 64 --num_warmup_epochs 5 --lr_scheduler_divide_every_n_epochs 30 --lr_scheduler_divisor 10 --num_epochs 100 \
  --use_nested_fsdp --use_gradient_checkpointing
```

**Results**: (pending)

#### Test ViT 10-billion model on v3-128

See https://github.com/ronghanghu/vit_10b_fsdp_example for test setup.

**Results**: (pending)

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+          "instead of using any of its submodules or its weights).")
+
+    super().__init__()
+    self.rank = xm.get_ordinal()
```
Thanks! I can try adding it in a future version. For this PR, I plan to first go with using all XLA devices as the whole group.

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1010 @@
+  def _reset_lazy_init(self) -> None:
+    """Reset instance so :func:`_lazy_init` will run on the next forward."""
+    self._is_root: Optional[bool] = None
+    self._output_pre_backward_hook_registered: Optional[List] = None
```
(changed from list to set)

ronghanghu commented on torch_xla/distributed/fsdp/xla_flatten_params_wrapper.py:
```diff
@@ -0,0 +1,586 @@
+    when different groups of params need different hyperparameters, like
+    learning rate or weight decay)
+
+  [1] https://github.com/SsnL/PyTorch-Reparam-Module
```
updated to cite FairScale's FlattenParamsWrapper

ronghanghu commented on torch_xla/distributed/fsdp/xla_flatten_params_wrapper.py:
```diff
@@ -0,0 +1,586 @@
+      module: nn.Module,
+      param_list: ParamGroups = None,
+      flat_param_names: Optional[List[str]] = None,
+      auto_unflatten_state_dict: Optional[bool] = True,
```
added doc for auto_unflatten_state_dict

ronghanghu commented on torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:
```diff
@@ -0,0 +1,1360 @@
+    .. note:: Gradient accumulation can be done without this context,
+        avoiding the extra TPU memory overhead, but with the extra
+        networking overhead.
+    """
```
Thanks! I can do more profiling tests later when https://github.com/pytorch/xla/issues/3446 is resolved.

JackCaoG commented:
build error should be resolve after a rebase.

ronghanghu commented:
(rebased onto master branch)

ronghanghu commented:
Update: our internal scaling tests show that we can scale up to transformer models with around 60 billion parameters on v3-512 using this FSDP PR. Then we encounter the issue of https://github.com/pytorch/xla/issues/3453 and cannot scale it to larger model sizes.

```
Computation requires more parameters (4748) than supported (limit 3304).
```

JackCaoG commented:
This is exciting! Thanks @ronghanghu !

ronghanghu commented:
> This is awesome @ronghanghu! Thank you! Do you plan to submit a separate ViT PR with your changes?

@miladm I think this PR should be ready for merge in its current shape once our internal accuracy tests pass. So far the loss curves of these accuracy tests look good but I'm still waiting to confirm their final results (should know by tomorrow). The current  PR commits above are exactly what we used internally at Meta AI in our scaling runs (including our ViT and other experiments).

Regarding the ViT scaling example, I would like to keep it as a separate repository in https://github.com/ronghanghu/vit_10b_fsdp_example for now because 1) it involves many open research questions on how to best train ViTs at this 10~60 billion parameter size (such as what optimizer to use, what learning rate to use, how to avoid training divergence) where we don't have an optimal recipe like the ImageNet or MNIST examples above yet, 2) it's more like an example of downstream application rather than features of PyTorch XLA itself, and 3) it relies on external dependencies such as `timm` to build the models.

So I think it's better to have a separate repository for scaling examples on top of PyTorch XLA as opposed to having them as part of PyTorch XLA itself. This is similar to keeping a separate [T5X repository](https://github.com/google-research/t5x) on top of JAX.

miladm commented:
@ronghanghu Feel free to merge the PR.

miladm reviewed (approved):
LGTM. Merging the PR.

hjm-aws commented:
@ronghanghu Hi Ronghang, we (AWS) plan to add ZeRO-2 optimization into your FSDP implementation. ZeRO-2 is just ZeRO-3 minus parameter sharding. We plan to add a flag to the constructor, and when the flag is `True`, we'll just skip the parameter sharding during init and skip all-gather before forward/backward. What do you think?

ronghanghu commented:
Hi @hjm-aws, you can already do ZeRO-2 by wrapping the entire model with a single (instead of nested) FSDP class with `reshard_after_forward=False`, and then applying the optimizer on the FSDP model's `.parameters()`. In this way, the full parameters will be all-gathered for the entire model before the forward pass and the gradients will be distributed via reduce-scatter.

This is actually the recommended way of implementing ZeRO-2 using the newer FSDP class (which is more convenient than putting the all-gather op in a separate optimizer class that handles sharded gradient update and parameter all-gathering/broadcasting)/

The FAIRScale library originally introduced a separate class `fairscale.nn.ShardedDataParallel` for ZeRO-2 as in https://fairscale.readthedocs.io/en/latest/deep_dive/oss_sdp_fsdp.html. But then it has been sort of obsolete as the newer FSDP class `fairscale.nn.FullyShardedDataParallel` can implement both ZeRO-2 and ZeRO-3 above.

ronghanghu commented:
@hjm-aws Also, regarding your earlier comment on adding process groups to allow mixed data parallelism and model parallelism -- this is something we are also planning to look at in Meta AI. So we could introduce a "group" option to this XLA FSDP in a future update after we first test it ourselves for mixed data + model parallelism

hjm-aws commented:
@ronghanghu Hi Ronghang, we need to avoid the all-gather that happens before each forward. Do you have a suggestion for achieving that?

ronghanghu commented:
> @ronghanghu Hi Ronghang, we need to avoid the all-gather that happens before each forward. Do you have a suggestion for achieving that?

Based on my understanding of ZeRO-2, I think an all-gather before the forward pass (or equivalently, after the sharded optimizer update of the previous iteration) is required in ZeRO-2. One needs to either put it in e.g. a sharded optimizer class that updates the parameter using sharded gradients and then broadcast the sharded updates via all-gather (as in SDP) or put it into the model before its forward pass (as in FSDP).

In the FSDP case, the model holds the sharded parameters in its idle state and only holds the full parameter during its forward pass, while in the SDP case the model holds the full parameter throughout all the time and only generates sharded parameters during its sharded optimizer update.
